Unable to use an existing network and `managedAPIServerLoadBalancer: true`

/kind bug

**What steps did you take and what happened:**

I created a cluster using:
 - an existing OpenStack network/subnet
 - `managedAPIServerLoadBalancer` enabled

The manifest I use is:

```
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
kind: OpenStackCluster
metadata:
  name: openstack
  namespace: default
spec:
  apiServerLoadBalancerAdditionalPorts:
  - 22
  apiServerLoadBalancerFloatingIP: 192.168.50.180
  apiServerLoadBalancerPort: 6443
  cloudName: openstack
  cloudsSecret:
    name: openstack-cloud-config
    namespace: default
  disablePortSecurity: true
  disableServerTags: true
  dnsNameservers:
  - 192.168.50.254
  externalNetworkId: dd02a459-0e4a-444d-9610-355709acbf1a
  managedAPIServerLoadBalancer: true
  network:
    name: "private-net"
  subnet:
    name: "private-subnet"
  useOctavia: false
```

`192.168.50.180` is an existing _unassigned_ floating address. And `private-net/subnet` are existing Network/Subnet.

Then no load balancer are created and thus the control plane node cannot initialize.

**What did you expect to happen:**

That a load balancer is created and that the control plan is initialized.

**Anything else you would like to add:**

If I set nodeCIDR instead of specifying and existing network and subnet. The setup works properly and a load balancer is created.

It seems that https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/controllers/openstackcluster_controller.go#L251 is missing a similar instruction `reconcileLoadBalancer` as the `else` case.

Note that if I delete the Cluster object then the `capo-controller-manager` crashes with:

```
E0420 12:06:03.224464       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 396 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1524ee0, 0x2427830)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1524ee0, 0x2427830)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
sigs.k8s.io/cluster-api-provider-openstack/pkg/cloud/services/loadbalancer.(*Service).DeleteLoadBalancerMember(0xc0006b2980, 0xc0005d21e0, 0x11, 0xc0004e2fc0, 0xc0004e2b40, 0xc0003a2500, 0x0, 0x0)
	/workspace/pkg/cloud/services/loadbalancer/loadbalancer.go:420 +0x1dd
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).reconcileDelete(0xc0003f6ba0, 0x191a680, 0xc000080080, 0x19292e0, 0xc000724330, 0xc000526230, 0xc0004e2fc0, 0xc0004e2b40, 0xc0006a7980, 0xc0003a2500, ...)
	/workspace/controllers/openstackmachine_controller.go:216 +0x759
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).Reconcile(0xc0003f6ba0, 0xc0005061f9, 0x7, 0xc00047e900, 0x1d, 0xc0004ddc00, 0x0, 0x0, 0x0)
	/workspace/controllers/openstackmachine_controller.go:146 +0x9e5
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0002220c0, 0x15780e0, 0xc0001c8620, 0xc000630500)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.1/pkg/internal/controller/controller.go:256 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0002220c0, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.1/pkg/internal/controller/controller.go:232 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0002220c0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.1/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0005e13b0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0005e13b0, 0x3b9aca00, 0x0, 0x1, 0xc000602000)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc0005e13b0, 0x3b9aca00, 0xc000602000)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.1/pkg/internal/controller/controller.go:193 +0x328
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x13c7a3d]
```

**Environment:**

- Cluster-api version: 

```
$ clusterctl version
clusterctl version: &version.Info{Major:"0", Minor:"3", GitVersion:"v0.3.3", GitCommit:"ecff70af1b839c4086335234d88b1b8c00c3383c", GitTreeState:"clean", BuildDate:"2020-03-27T22:25:19Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
```

- Minikube/KIND version:

```
$ kind version
kind v0.7.0 go1.13.6 linux/amd64
```

- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

--- a/controllers/openstackcluster_controller.go
+++ b/controllers/openstackcluster_controller.go
@@ -160,7 +160,8 @@ func (r *OpenStackClusterReconciler) reconcileDelete(ctx context.Context, log lo
 		log.Info("OpenStack router deleted successfully")
 	}
 
-	if openStackCluster.Status.Network != nil {
+	// if NodeCIDR was not set, no network was created.
+	if openStackCluster.Status.Network != nil && openStackCluster.Spec.NodeCIDR != "" {
 		log.Info("Deleting network", "name", openStackCluster.Status.Network.Name)
 		if err := networkingService.DeleteNetwork(openStackCluster.Status.Network); err != nil {
 			return ctrl.Result{}, errors.Errorf("failed to delete network: %v", err)
@@ -334,11 +335,12 @@ func (r *OpenStackClusterReconciler) reconcileNetworkComponents(log logr.Logger,
 		if err != nil {
 			return errors.Errorf("failed to reconcile router: %v", err)
 		}
-		if openStackCluster.Spec.ManagedAPIServerLoadBalancer {
-			err = loadBalancerService.ReconcileLoadBalancer(clusterName, openStackCluster)
-			if err != nil {
-				return errors.Errorf("failed to reconcile load balancer: %v", err)
-			}
+	}
+
+	if openStackCluster.Spec.ManagedAPIServerLoadBalancer {
+		err = loadBalancerService.ReconcileLoadBalancer(clusterName, openStackCluster)
+		if err != nil {
+			return errors.Errorf("failed to reconcile load balancer: %v", err)
 		}
 	}
 

--- a/controllers/openstackcluster_controller.go
+++ b/controllers/openstackcluster_controller.go
@@ -21,8 +21,10 @@ import (
 	"fmt"
 
 	"github.com/go-logr/logr"
+	"github.com/gophercloud/gophercloud"
 	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
 	"github.com/gophercloud/gophercloud/openstack/networking/v2/subnets"
+	"github.com/gophercloud/utils/openstack/clientconfig"
 	"github.com/pkg/errors"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/client-go/tools/record"
@@ -189,8 +191,6 @@ func contains(arr []string, target string) bool {
 func (r *OpenStackClusterReconciler) reconcileNormal(ctx context.Context, log logr.Logger, patchHelper *patch.Helper, cluster *clusterv1.Cluster, openStackCluster *infrav1.OpenStackCluster) (ctrl.Result, error) {
 	log.Info("Reconciling Cluster")
 
-	clusterName := fmt.Sprintf("%s-%s", cluster.Namespace, cluster.Name)
-
 	// If the OpenStackCluster doesn't have our finalizer, add it.
 	controllerutil.AddFinalizer(openStackCluster, infrav1.ClusterFinalizer)
 	// Register the finalizer immediately to avoid orphaning OpenStack resources on delete
@@ -208,71 +208,9 @@ func (r *OpenStackClusterReconciler) reconcileNormal(ctx context.Context, log lo
 		return reconcile.Result{}, err
 	}
 
-	networkingService, err := networking.NewService(osProviderClient, clientOpts, log)
+	err = r.reconcileNetworkComponents(log, osProviderClient, clientOpts, cluster, openStackCluster)
 	if err != nil {
 		return reconcile.Result{}, err
-	}
-
-	loadBalancerService, err := loadbalancer.NewService(osProviderClient, clientOpts, log, openStackCluster.Spec.UseOctavia)
-	if err != nil {
-		return reconcile.Result{}, err
-	}
-
-	log.Info("Reconciling network components")
-	if openStackCluster.Spec.NodeCIDR == "" {
-		log.V(4).Info("No need to reconcile network, searching network and subnet instead")
-
-		netOpts := networks.ListOpts(openStackCluster.Spec.Network)
-		networkList, err := networkingService.GetNetworksByFilter(&netOpts)
-		if err != nil && len(networkList) == 0 {
-			return reconcile.Result{}, errors.Errorf("failed to find network: %v", err)
-		}
-		if len(networkList) > 1 {
-			return reconcile.Result{}, errors.Errorf("failed to find only one network (result: %v): %v", networkList, err)
-		}
-		openStackCluster.Status.Network = &infrav1.Network{
-			ID:   networkList[0].ID,
-			Name: networkList[0].Name,
-		}
-
-		subnetOpts := subnets.ListOpts(openStackCluster.Spec.Subnet)
-		subnetOpts.NetworkID = networkList[0].ID
-		subnetList, err := networkingService.GetSubnetsByFilter(&subnetOpts)
-		if err != nil || len(subnetList) == 0 {
-			return reconcile.Result{}, errors.Errorf("failed to find subnet: %v", err)
-		}
-		if len(subnetList) > 1 {
-			return reconcile.Result{}, errors.Errorf("failed to find only one subnet (result: %v): %v", subnetList, err)
-		}
-		openStackCluster.Status.Network.Subnet = &infrav1.Subnet{
-			ID:   subnetList[0].ID,
-			Name: subnetList[0].Name,
-			CIDR: subnetList[0].CIDR,
-		}
-	} else {
-		err := networkingService.ReconcileNetwork(clusterName, openStackCluster)
-		if err != nil {
-			return reconcile.Result{}, errors.Errorf("failed to reconcile network: %v", err)
-		}
-		err = networkingService.ReconcileSubnet(clusterName, openStackCluster)
-		if err != nil {
-			return reconcile.Result{}, errors.Errorf("failed to reconcile subnets: %v", err)
-		}
-		err = networkingService.ReconcileRouter(clusterName, openStackCluster)
-		if err != nil {
-			return reconcile.Result{}, errors.Errorf("failed to reconcile router: %v", err)
-		}
-		if openStackCluster.Spec.ManagedAPIServerLoadBalancer {
-			err = loadBalancerService.ReconcileLoadBalancer(clusterName, openStackCluster)
-			if err != nil {
-				return reconcile.Result{}, errors.Errorf("failed to reconcile load balancer: %v", err)
-			}
-		}
-	}
-
-	err = networkingService.ReconcileSecurityGroups(clusterName, openStackCluster)
-	if err != nil {
-		return reconcile.Result{}, errors.Errorf("failed to reconcile security groups: %v", err)
 	}
 
 	// Set APIEndpoints so the Cluster API Cluster Controller can pull them
@@ -337,6 +275,79 @@ func (r *OpenStackClusterReconciler) reconcileNormal(ctx context.Context, log lo
 	openStackCluster.Status.Ready = true
 	log.Info("Reconciled Cluster create successfully")
 	return ctrl.Result{}, nil
+}
+
+func (r *OpenStackClusterReconciler) reconcileNetworkComponents(log logr.Logger, osProviderClient *gophercloud.ProviderClient, clientOpts *clientconfig.ClientOpts, cluster *clusterv1.Cluster, openStackCluster *infrav1.OpenStackCluster) error {
+	clusterName := fmt.Sprintf("%s-%s", cluster.Namespace, cluster.Name)
+
+	networkingService, err := networking.NewService(osProviderClient, clientOpts, log)
+	if err != nil {
+		return err
+	}
+
+	loadBalancerService, err := loadbalancer.NewService(osProviderClient, clientOpts, log, openStackCluster.Spec.UseOctavia)
+	if err != nil {
+		return err
+	}
+
+	log.Info("Reconciling network components")
+	if openStackCluster.Spec.NodeCIDR == "" {
+		log.V(4).Info("No need to reconcile network, searching network and subnet instead")
+
+		netOpts := networks.ListOpts(openStackCluster.Spec.Network)
+		networkList, err := networkingService.GetNetworksByFilter(&netOpts)
+		if err != nil && len(networkList) == 0 {
+			return errors.Errorf("failed to find network: %v", err)
+		}
+		if len(networkList) > 1 {
+			return errors.Errorf("failed to find only one network (result: %v): %v", networkList, err)
+		}
+		openStackCluster.Status.Network = &infrav1.Network{
+			ID:   networkList[0].ID,
+			Name: networkList[0].Name,
+		}
+
+		subnetOpts := subnets.ListOpts(openStackCluster.Spec.Subnet)
+		subnetOpts.NetworkID = networkList[0].ID
+		subnetList, err := networkingService.GetSubnetsByFilter(&subnetOpts)
+		if err != nil || len(subnetList) == 0 {
+			return errors.Errorf("failed to find subnet: %v", err)
+		}
+		if len(subnetList) > 1 {
+			return errors.Errorf("failed to find only one subnet (result: %v): %v", subnetList, err)
+		}
+		openStackCluster.Status.Network.Subnet = &infrav1.Subnet{
+			ID:   subnetList[0].ID,
+			Name: subnetList[0].Name,
+			CIDR: subnetList[0].CIDR,
+		}
+	} else {
+		err := networkingService.ReconcileNetwork(clusterName, openStackCluster)
+		if err != nil {
+			return errors.Errorf("failed to reconcile network: %v", err)
+		}
+		err = networkingService.ReconcileSubnet(clusterName, openStackCluster)
+		if err != nil {
+			return errors.Errorf("failed to reconcile subnets: %v", err)
+		}
+		err = networkingService.ReconcileRouter(clusterName, openStackCluster)
+		if err != nil {
+			return errors.Errorf("failed to reconcile router: %v", err)
+		}
+		if openStackCluster.Spec.ManagedAPIServerLoadBalancer {
+			err = loadBalancerService.ReconcileLoadBalancer(clusterName, openStackCluster)
+			if err != nil {
+				return errors.Errorf("failed to reconcile load balancer: %v", err)
+			}
+		}
+	}
+
+	err = networkingService.ReconcileSecurityGroups(clusterName, openStackCluster)
+	if err != nil {
+		return errors.Errorf("failed to reconcile security groups: %v", err)
+	}
+
+	return nil
 }
 
 func (r *OpenStackClusterReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {

